### ISSUE TITLE
[NEW] Rethink and Redesign the MEET Protocol to Address Increasing Brittleness and Complexity

### ISSUE THREAD

**Issue by PingXie** [madolson/placeholderkv#1471]
This code flow has become too brittle and we are always in catchup mode. I think we should step back and reconsider the `MEET` flow. 

@madolson / @PingXie / @enjoy-binbin

_Originally posted by @hpatro in https://github.com/valkey-io/valkey/pull/1436#pullrequestreview-2502967525_

**Comment by madolson** [comment]
I'm not sure what you're suggesting. If we move to a single source of truth, then we would have a different meet system and would get consensus through the other library we pick. 

The current meet approach just didn't handle partial timeouts, which is part of the price for pay for building a bespoke consensus algorithm.

**Comment by PingXie** [comment]
> If we move to a single source of truth,

Are you talking about cluster v2? If so, I don't think we should tie the improvement of existing cluster to the timeline of the v2 project. I think that we at least need to document the current meet behavior and then see if we can reason about future patches more easily. "Redesign" though does sound like the wrong word. Let me update the title to clarify the issue better.

**Comment by madolson** [comment]
> Are you talking about cluster v2? If so, I don't think we should tie the improvement of existing cluster to the timeline of the v2 project.

Yeah. 

> I think that we at least need to document the current meet behavior and then see if we can reason about future patches more easily. "Redesign" though does sound like the wrong word. Let me update the title to clarify the issue better.

I'm always here for better documentation. The more I think about cluster v2 the more I think we should really just take a fresh pass at the current cluster code and clean it up and try to simplify it. There are many undocumented assumptions, confusing code paths, and the algorithms that were used are brittle in failure modes. Some fuzzy/chaos testing would also might be a good idea.

**Comment by enjoy-binbin** [comment]
What I want to add to MEET is to add some auth meet / auth handshake. We often suffer from some cluster merging problems, so in internal we had meet auth and handshake auth to control the CLUSTER MEET and the clusterAddNode(in the gossip one). Our current code relies too much on CLUSTER MEET / clusterAddNode. If the administrator enters the wrong IP/port, or there are some dead nodes in the cluster, the cluster will be in chaos when the IP/port is reused.

**Comment by madolson** [comment]
> What I want to add to MEET is to add some auth meet / auth handshake. We often suffer from some cluster merging problems, so in internal we had meet auth and handshake auth to control the CLUSTER MEET and the clusterAddNode(in the gossip one). Our current code relies too much on CLUSTER MEET / clusterAddNode. If the administrator enters the wrong IP/port, or there are some dead nodes in the cluster, the cluster will be in chaos when the IP/port is reused.

In AWS talked about having a config to isolate a node, so that it would just silently drop all meet and pong packets and wouldn't try to communicate with a cluster. Having a unique cluster secret that is shared between all the nodes in a cluster also makes a lot of sense. We could add that with a ping extension I think?

**Comment by enjoy-binbin** [comment]
> Having a unique cluster secret that is shared between all the nodes in a cluster also makes a lot of sense. We could add that with a ping extension I think?

yes, this can be fit into ping extension. so a unique cluster secret is added into the packet, when parsing the PING/PONG/MEET, we check it in clusterIsValidPacket, if the secret does not match, we drop the packet? The secret should be small so the overhead is ok i think. This seem ok for me. I can try to do it in the next few weeks.

**Comment by enjoy-binbin** [comment]
I forgot to mention this. The first MEET packet is not able to carry the extensions because of #52, so we can not drop the packet in the very beginning (the first MEET packet). @hpatro do you have some thoughts?

**Comment by hpatro** [comment]
> I forgot to mention this. The first MEET packet is not able to carry the extensions because of [#52](https://github.com/valkey-io/valkey/pull/52), so we can not drop the packet in the very beginning (the first MEET packet). [@hpatro](https://github.com/hpatro) do you have some thoughts?

Hmmmm, that was intentional to avoid sending additional payload in the packet which can't be handled by an older node which didn't support extensions. That makes your above design complex. 
One possibility I can think of is enabling this extension based on certain server side config and would only be enabled in a cluster where all nodes are guaranteed to handle extensions.

**Comment by enjoy-binbin** [comment]
> One possibility I can think of is enabling this extension based on certain server side config and would only be enabled in a cluster where all nodes are guaranteed to handle extensions.

This seems to be the only way I can think of right now too. We add extensions in 7.0, maybe in a future version we can always send the extension without checking for old 6.2 versions...

**Comment by hpatro** [comment]
> > One possibility I can think of is enabling this extension based on certain server side config and would only be enabled in a cluster where all nodes are guaranteed to handle extensions.
> 
> This seems to be the only way I can think of right now too. We add extensions in 7.0, maybe in a future version we can always send the extension without checking for old 6.2 versions...

Possibly. However, I sense we will have folks migrating from Redis for a long period of time. So, we need to keep supporting the migration use cases.
